etcdserver/etcdhttp/client.go

```diff
@@ -114,6 +128,11 @@ func (h *keysHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		writeError(w, err)
 		return
 	}
+	// The path must be valid at this point (we've parsed the request successfully).
+	if !hasKeyPrefixAccess(h.sec, r, r.URL.Path[len(keysPrefix):]) {
```

Comment:
The reason I did it this way (also valid on *Store, but a little strange) is that if the store is nil, then it can respond appropriately